### PR TITLE
Tighter check when validating tables with links

### LINKED ISSUE #1114: Check if migration is needed is too loose for RealmList
If a model class has a `RealmList<A>` field is change to a `RealmList<B>`, the proxy class' `validateTable` will not throw a `MigrationIsNeeded` exception. This can lead to an inconsistent Realm file as discovered in #1094.

### PR DESCRIPTION
**Pull request by kneth**
This is fixes #1114 (originally reported in #1094).

@emanuelez @cmelchior @bmunkholm

### CODE REVIEW

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 1]
```diff
@@ -379,21 +379,38 @@ private void emitValidateTableMethod(JavaWriter writer) throws IOException {
...
         writer.emitStatement("columnTypes.put(table.getColumnName(i), table.getColumnType(i))");
         writer.endControlFlow();
 
-        // For each field verify there is a corresponding column
+        // Populate column indices
+        writer.emitEmptyLine();
+        writer.emitStatement("columnIndices = new HashMap<String, Long>()");
+        writer
+                .beginControlFlow("for (String fieldName : getFieldNames())")
+                .emitStatement("long index = table.getColumnIndex(fieldName)")
```
Comment (emanuelez) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: Use indentation to improve readability

File: changelog.txt [hunk 2]
```diff
@@ -1,7 +1,8 @@
 0.81.0
  * Fixed bug where Realm.createOrUpdateWithJson() reset Date and Binary data to default values if not found in the JSON output.
  * RealmBaseAdapter now allow RealmResults to be null (thanks @zaki50).
-
+ * Fixing bug where a change to a model class (RealmList<A> to RealmList<B>) will throw an MigrationNeeded exception.
```
Comment (cmelchior) [inline changelog.txt]: RealmMigrationNeededException

Comment (emanuelez) [inline changelog.txt]: Fixed a bug in which a change to a model class (RealmList<A> to RealmList<B>) would not throw a MigrationNeededException.

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 3]
```diff
@@ -435,18 +452,17 @@ private void emitValidateTableMethod(JavaWriter writer) throws IOException {
...
                 writer.endControlFlow();
-                // TODO: Replace with a proper comparison
-//                writer.emitStatement("Table table_%d = transaction.getTable(\"%s%s\")", columnNumber, TABLE_PREFIX, fieldTypeName);
-//                writer.beginControlFlow("if (table.getLinkTarget(%d).equals(table_%d))", columnNumber, columnNumber);
-//                writer.emitStatement("throw new IllegalStateException(\"Mismatching link tables for column '%s'\")",
-//                        fieldName);
-//                writer.endControlFlow();
+                writer.emitStatement("Table table_%d = transaction.getTable(\"%s%s\")", fieldIndex, Constants.TABLE_PREFIX, fieldTypeSimpleName);
+                writer.beginControlFlow("if (!table.getLinkTarget(%s).equals(table_%d))", staticFieldIndexVarName(field), fieldIndex);
+                writer.emitStatement("throw new IllegalStateException(\"Mismatching link tables for column '%s'\")",
```
Comment (cmelchior) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: Too core centric. Rephrase to match user world view. Perhaps "Invalid RealmList type. Was x, expected y"

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 4]
```diff
@@ -435,18 +452,17 @@ private void emitValidateTableMethod(JavaWriter writer) throws IOException {
                 writer.emitStatement("throw new IllegalStateException(\"Missing table '%s%s' for column '%s'\")",
                         Constants.TABLE_PREFIX, fieldTypeSimpleName, fieldName);
                 writer.endControlFlow();
-                // TODO: Replace with a proper comparison
```
Comment (cmelchior) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: Yay, finally getting rid of those :)

File: realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java [hunk 5]
```diff
@@ -261,33 +282,23 @@ public static void validateTable(ImplicitTransaction transaction) {
             if (!transaction.hasTable("class_AllTypes")) {
                 throw new IllegalStateException("Missing table 'class_AllTypes' for column 'columnObject'");
             }
-            if(!columnTypes.containsKey("columnRealmList")) {
+            Table table_7 = transaction.getTable("class_AllTypes");
+            if (!table.getLinkTarget(INDEX_COLUMNOBJECT).equals(table_7)) {
```
Comment (cmelchior) [inline realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Isn't it dangerous to test for type validity before checking if the field exists at all?

File: realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java [hunk 6]
```diff
@@ -195,15 +195,36 @@ public static Table initTable(ImplicitTransaction transaction) {
     }
 
     public static void validateTable(ImplicitTransaction transaction) {
-        if(transaction.hasTable("class_AllTypes")) {
+        if (transaction.hasTable("class_AllTypes")) {
             Table table = transaction.getTable("class_AllTypes");
-            if(table.getColumnCount() != 9) {
+
+            if (table.getColumnCount() != 9) {
                 throw new IllegalStateException("Column count does not match");
```
Comment (cmelchior) [inline realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Perhaps rephrase to "Number of fields does not match the expected. Is x, expected y"?

File: realm/src/androidTest/java/io/realm/entities/CatOwner.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm.entities;
+
+
+import io.realm.RealmList;
+import io.realm.RealmObject;
+
+public class CatOwner extends RealmObject {
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/entities/CatOwner.java]: Is it really needed to add a new entity for this. Can't we reuse some of the older ones?

File: realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java (hunk 6 above)
Comment (bmunkholm) [inline realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: This is a generated class - so it should be fixed in the generator.

File: realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java (hunk 5 above)
Comment (bmunkholm) [inline realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: same

File: realm/src/androidTest/java/io/realm/entities/CatOwner.java (hunk 7 above)
Comment (kneth) [inline realm/src/androidTest/java/io/realm/entities/CatOwner.java]: No.

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 8]
```diff
@@ -387,7 +387,8 @@ private void emitValidateTableMethod(JavaWriter writer) throws IOException {
 
         // verify number of columns
         writer.beginControlFlow("if (table.getColumnCount() != " + metadata.getFields().size() + ")");
-        writer.emitStatement("throw new RealmMigrationNeededException(transaction.getPath(), \"Field count does not match\")");
+        writer.emitStatement("throw new RealmMigrationNeededException(transaction.getPath(), \"Field count does not match - expected %d\")",
```
Comment (cmelchior) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: Why not write the current table count as well? Eg. Field count does not match. Was x, expected y? Then you don't have to attach a debugger afterwards to check it :)

File: realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java [hunk 9]
```diff
@@ -263,33 +284,24 @@ public static void validateTable(ImplicitTransaction transaction) {
             if (!transaction.hasTable("class_AllTypes")) {
                 throw new RealmMigrationNeededException(transaction.getPath(), "Missing class 'class_AllTypes' for field 'columnObject'");
             }
-            if(!columnTypes.containsKey("columnRealmList")) {
+
+            Table table_7 = transaction.getTable("class_AllTypes");
+            if (!table.getLinkTarget(INDEX_COLUMNOBJECT).equals(table_7)) {
+                throw new RealmMigrationNeededException(transaction.getPath(), "Mismatching link tables for field 'columnObject'");
```
Comment (cmelchior) [inline realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Very core centric error msg. Perhaps "Mismatching field type for field 'xxx'...Even better would be "Mismatching field type for field 'xxx'. Was y expected z". y and z probably has the format "class_Y" and "class_Z" but that is ok I think

File: realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java [hunk 10]
```diff
@@ -263,33 +284,24 @@ public static void validateTable(ImplicitTransaction transaction) {
...
-            columnIndices = new HashMap<String, Long>();
-            for (String fieldName : getFieldNames()) {
-                long index = table.getColumnIndex(fieldName);
-                if (index == -1) {
-                    throw new RealmMigrationNeededException(transaction.getPath(), "Field '" + fieldName + "' not found for type AllTypes");
-                }
-                columnIndices.put(fieldName, index);
+            Table table_8 = transaction.getTable("class_AllTypes");
+            if (!table.getLinkTarget(INDEX_COLUMNREALMLIST).equals(table_8)) {
+                throw new RealmMigrationNeededException(transaction.getPath(), "Mismatching link list tables for column 'columnRealmList'");
```
Comment (cmelchior) [inline realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Same as above

File: changelog.txt [hunk 11]
```diff
@@ -2,7 +2,8 @@
  * Fixed bug where Realm.createOrUpdateWithJson() reset Date and Binary data to default values if not found in the JSON output.
  * Fixed a memory leak when using RealmBaseAdapter.
  * RealmBaseAdapter now allow RealmResults to be null (thanks @zaki50).
-
+ * Fixed a bug which a change to a model class (RealmList<A> to RealmList<B>) would not throw a RealmMigrationNeededException.
```
Comment (cmelchior) [inline changelog.txt]: which -> where

**Comment by cmelchior** [comment]
Apart from comments :+1:

**Comment by emanuelez** [comment]
I think we're good now :+1:

**Pull request merged**